server/data/data-library/src/main/java/org/apache/james/user/lib/model/Algorithm.java

```diff
@@ -19,19 +19,57 @@
 
 package org.apache.james.user.lib.model;
 
+import java.util.Arrays;
 import java.util.Objects;
 
+import com.google.common.collect.ImmutableList;
+
 public class Algorithm {
-    private static final boolean LEGACY = true;
-    public static final String LEGACY_SUFFIX = "-legacy";
 
-    public static Algorithm of(String rawValue) {
-        if (rawValue.endsWith(LEGACY_SUFFIX)) {
-            return new Algorithm(rawValue, rawValue.substring(0, rawValue.length() - LEGACY_SUFFIX.length()), LEGACY);
+    public enum HashingMode {
+        LEGACY(LEGACY_FACTORY),
+        DEFAULT(DEFAULT_FACTORY);
+
+        public static HashingMode parse(String value) {
+            return Arrays.stream(values())
+                .filter(aValue -> value.equalsIgnoreCase(aValue.toString()))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Unsuported value for HashingMode: " + value + ". Should be one of " + ImmutableList.copyOf(values())));
+        }
+
+        private final Factory factory;
+
+        HashingMode(Factory factory) {
+            this.factory = factory;
+        }
+
+        public Factory getFactory() {
+            return factory;
+        }
+    }
+
+    public interface Factory {
+        Algorithm of(String rawValue);
+    }
+
+    public static class LegacyFactory implements Factory {
+        @Override
+        public Algorithm of(String rawValue) {
+            return new Algorithm(rawValue, rawValue, LEGACY);
```

Comment:
If the two arguments for Algorithm `rawValue` and `algorithmName` are the same now, it's probably better to drop one?